torch/csrc/autograd/input_metadata.h

```diff
@@ -67,9 +104,33 @@ struct InputMetadata {
     return at::zeros(shape_, options_);
   }
 
+  at::IntArrayRef compute_shape(const at::Tensor& t){
+    if(t.is_nested()){
+      return at::IntArrayRef{};
+    }
+    return t.sizes();
+  }
+
+  c10::optional<const at::Tensor> compute_nested_shape_from_tensor(const at::Tensor& t){
+    if(t.is_nested()){
+      auto nested_impl = at::native::get_nested_tensor_impl(t);
+      return nested_impl->get_nested_size_tensor();
+    }
+    return c10::optional<const at::Tensor> {};
+  }
+
+  bool is_same_shape(const at::Tensor& grad) const {
+    if (is_nested_tensor()) {
+      return grad.nested_size_tensor().equal(nested_shape());
+    } else {
+      return grad.sizes().equals(shape());
+    }
+  }
+
 private:
   const at::TensorOptions options_;
   at::DimVector shape_;
+  c10::optional<const at::Tensor> nested_shape_;
```

Comment:
I suggest we use optional<DimVector>

Reply:
So when we say we can re-use the dim vector is this what we have in mind. 
- add a  boolean field to the struct `is_nested()`
- And to use a concrete example lets say we have a batch of N tensors all 2 dimensional. The padded tensor would be N X L X D.  Currently the size information for a nested tensor is stored represented by a tensor. In this case it would be 3 dimensional.
```Python
tensor([
        [2,64],
        [8, 64],
        [12, 64],
        [5, 64],
        .
        .
        .
       ])
```
if we were to store this information in a DimVector that would look like this:
``` Python
[2, 64, 8, 64, 12, 64, 5, 64, ... ]
```
If we needed to reconstruct the size we could use another `optional<size_t> inner_dim_size` which would be 2 (the first stride of the nested_size_tensor).  

Since meta data is mostly about comparisons  between shapes and not absolute shapes we might not need this extra member.  Unless we want to check if grad is expandable to metadata shape.

I have two questions from this though: 
1. We have essentially broken up the compact representation of a Tensor into a buffer and a stride. Are the advantages of doing this better than storing an optional Tensor which itself is a pointer.

2. For batch sizes where the overhead of nested tensor make sense (lets say 256 or greater in this scenario) is DimVector the right data structure to store this information? Crucially we would not be able to de-dupe the repeated `64` in this case, so the dim vector would be of size 256*2 = 512 .

Reply:
Wouldn't the tensor be a pointer to a pointer? Storage itself is still another indirection away from tensor, so it isn't that compact. 

When your list gets that large you probably aren't getting any extra utility from using a DimVector, you'll probably always do a heap allocation, but that is what a normal vector does anyway, so not sure if there is a downside to using DimVector.